Add ERC: Universal Orchestrator RPC (#778)
ethereum/ERCs: ERCS/erc-7845.md

---
eip: 7845

> title: Minimal Orchestrator RPC

Author: Understood - the name has been changed to Universal Orchestrator RPC. Just a note, i'm currently doing research to figure out if the term Orchestrator is still current.

description: A minimal data model that an Orchestrator can understand to help fulfil a request for a solution from a wallet or any other system.
author: Kieran Goodary (@IAmKio), Pillar Wallet (@pillarwallet), Luke Wickens (@lbw33), Rana Khoury (@RanaBug)
discussions-to: https://ethereum-magicians.org/t/erc-7845-minimal-orchestrator-rpc/21885
status: Draft
type: Standards Track
category: ERC
created: 2024-11-07
---

## Abstract

> > "Hey Google, swap my SHIB for Pillar"

Reviewer: We like to avoid mentioning specific products in proposals. It unfairly advertises projects that exist today over ones that come after going final.

Author: Agreed - this has been changed. Thanks.

## Motivation

Data model standards can be written in any shape. A system will often expose their external interface but require that the request to the aforementioned interface is modelled in a way that the service understands. This creates a huge level of inconsistency and in turn makes Orchestrator interoperability more difficult.

Orchestrators will become more widespread and numerous over time. This is especially true with the advent of Artificial Intelligence (AI) driven systems, the continued advancement of Human Computer Interaction (HCI) devices (especially those that are voice controlled) and the emergence of Extended Reality (XR) platforms.

Standardising the request object that an Orchestrator can understand from a wallet will drive adoption and make decentralised app development easier for developers that don't know how to make on-chain transactions or have the required technical understanding of block building systems.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> The Orchestrator's external interface(s) that expose functionality to an end-user application or another system `MUST` use JavaScript Object Notation (JSON).

Reviewer: Reserve backticks for inline code. The RFC 2119 key words should be unformatted uppercase.

> The following data definitions are available and `MUST` **prefer chain abstraction (ChA¹)**, unless stated.

Reviewer: It is unclear what "prefer chain abstraction" means.

Author: I have added an explanation for clarity.

> The following sequence diagram shows the flow that this ERC aims to standardise:

Reviewer: You should write as if the proposal was already Final. Maybe something like:

```suggestion
The following sequence diagram shows the flow of events in this proposal:
```

Author: Ahh understood - thanks, i'll check the rest of the document for this also.

![Sequence Diagram](../assets/eip-7845/sequence.png)

This specification follows the top level data shape of Ethereum JSON-RPC requests, as shown below:

### Request

The request definition is what a wallet sends to an Orchestrator for solutions to one or more problems. The request follows the specification from Ethereum JSON-RPC.

> #### RPC

Reviewer: This section seems to be re-explaining JSON RPC. I'm not sure how necessary it is.

Author: Whilst i understand what you're suggesting - for the sake of being clear and completion, i would prefer to leave this here to make it crystal clear that this is an Ethereum RPC request and that the `params` property is a `Problem[]`.

- It's understood that the Orchestrator interpretations and implementations will vary, so where possible the specification enforces `REQUIRED` and `MUST` to provide a minimal level of service to an end-user or another service.
- The specification is **NOT** intended to standardise or modify the internal data structure or communication layer of an Orchestrator.
- Other parameters that could be considered, such as gas limits and estimations, are delegated back to the wallet as ultimately it is the wallet that will execute the solution(s).

## Backwards Compatibility

No backward compatibility issues found.

## Reference Implementation

### Example requests and responses for solutions

The following examples show a few common scenarios with their requests to, and from, an Orchestrator. All examples are chain abstracted (ChA¹) by default, unless specified.

#### Sending an [ERC-20](./eip-20.md) token to another address

The following request performs an action:

- from 0xbafB4E1EFA94B359e2E175CF6156AedA2cACa165
- towards address 0x50840CE036eEf2005d3c4d6f6Eb65f8116a01629
  - with symbol USDC, amount 5

> :::info

Reviewer: I'm not sure this'll render the way you want. Double check it after merging.

Author: Sorry you're right i was using a reference from elsewhere. I have now changed it to the alert standard here: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts
